src/overmap.cpp

```diff
@@ -713,52 +712,52 @@ bool overmap::has_note(int const x, int const y, int const z) const
         return false;
     }
 
-    for (auto &i : layer[z + OVERMAP_DEPTH].notes) {
-        if (i.x == x && i.y == y) {
-            return true;
-        }
-    }
-    return false;
+    auto const &notes = layer[z + OVERMAP_DEPTH].notes;
+    return std::any_of(begin(notes), end(notes), [&](om_note const& n) {
+        return n.x == x && n.y == y;
+    });
 }
 
-std::string const &overmap::note(int const x, int const y, int const z) const
+std::string const& overmap::note(int const x, int const y, int const z) const
 {
+    static std::string const fallback {};
+
     if (z < -OVERMAP_DEPTH || z > OVERMAP_HEIGHT) {
-        return nullstr;
+        return fallback;
     }
 
-    for (auto &i : layer[z + OVERMAP_DEPTH].notes) {
-        if (i.x == x && i.y == y) {
-            return i.text;
-        }
-    }
+    auto const &notes = layer[z + OVERMAP_DEPTH].notes;
+    auto const it = std::find_if(begin(notes), end(notes), [&](om_note const& n) {
+        return n.x == x && n.y == y;
+    });
 
-    return nullstr;
+    return (it != std::end(notes)) ? it->text : fallback;
 }
 
-void overmap::add_note(int const x, int const y, int const z, std::string const &message)
+void overmap::add_note(int const x, int const y, int const z, std::string message)
 {
     if (z < -OVERMAP_DEPTH || z > OVERMAP_HEIGHT) {
         debugmsg("Attempting to add not to overmap for blank layer %d", z);
         return;
     }
 
-    for (size_t i = 0; i < layer[z + OVERMAP_DEPTH].notes.size(); i++) {
-        if (layer[z + OVERMAP_DEPTH].notes[i].x == x && layer[z + OVERMAP_DEPTH].notes[i].y == y) {
-            if (message.empty()) {
-                layer[z + OVERMAP_DEPTH].notes.erase(layer[z + OVERMAP_DEPTH].notes.begin() + i);
-            } else {
-                layer[z + OVERMAP_DEPTH].notes[i].text = message;
-            }
-            return;
-        }
-    }
-    if (message.length() > 0) {
-        layer[z + OVERMAP_DEPTH].notes.push_back(om_note(x, y, layer[z + OVERMAP_DEPTH].notes.size(),
-                message));
+    auto &notes = layer[z + OVERMAP_DEPTH].notes;
+    auto const it = std::find_if(begin(notes), end(notes), [&](om_note const& n) {
+        return n.x == x && n.y == y;
+    });
+
+    if (message.empty() && it != notes.end()) {
+        notes.erase(it);
+    } else if(!message.empty()) {
+        notes.emplace_back(om_note {std::move(message), x, y});
```

Comment:
Doesn't that change the logic (haven't tested) when this is called with a non-empty message on a place that already has a note? In that case `it != end`, but message is non-empty, so a *new* note is *added*, the previous code would instead change the existing entry.